// ionic4-tooltips/esm2015/main.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
/** Components **/
export * from './components/tooltip-box/tooltip-box.component';
import { TooltipBoxComponent } from './components/tooltip-box/tooltip-box.component';
/** Controllers **/
export * from './controllers/tooltip.cotroller';
import { TooltipController } from './controllers/tooltip.cotroller';
/** Directives **/
export * from './directives/tooltip.directive';
import { TooltipDirective } from './directives/tooltip.directive';
export class TooltipsModule {
    static forRoot() {
        return {
            ngModule: TooltipsModule,
            providers: [
                TooltipController
            ],
        };
    }
}
TooltipsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    TooltipDirective,
                    TooltipBoxComponent
                ],
                entryComponents: [
                    TooltipBoxComponent
                ],
                exports: [
                    TooltipDirective
                ],
                imports: [
                    CommonModule
                ]
            },] }
];
